// next@15.4.0-canary.59/dist/esm/shared/lib/devtool/app-segment-tree.js
'use client';
import { useEffect, useSyncExternalStore } from 'react';
import { createTrie } from './trie';
const DEFAULT_CLIENT_STATE = typeof window === 'undefined' ? undefined : createTrie({
    getKey: (item)=>item.pagePath
});
function getSegmentTreeClientState() {
    if (typeof window === 'undefined') {
        return {};
    }
    if (!window.__NEXT_DEVTOOLS_CLIENT_STATE) {
        window.__NEXT_DEVTOOLS_CLIENT_STATE = {
            // Initial state
            tree: DEFAULT_CLIENT_STATE
        };
    }
    return window.__NEXT_DEVTOOLS_CLIENT_STATE;
}
const listeners = typeof window === 'undefined' ? null : new Set();
const createSegmentTreeStore = ()=>{
    if (typeof window === 'undefined') {
        return {
            subscribe: ()=>()=>void 0,
            getSnapshot: ()=>({}),
            getServerSnapshot: ()=>undefined
        };
    }
    // return a store that can be used by useSyncExternalStore
    return {
        subscribe: (callback)=>{
            listeners == null ? void 0 : listeners.add(callback);
            return ()=>listeners == null ? void 0 : listeners.delete(callback);
        },
        getSnapshot: ()=>{
            return getSegmentTreeClientState();
        },
        getServerSnapshot: ()=>{
            return undefined;
        }
    };
};
const { subscribe, getSnapshot, getServerSnapshot } = createSegmentTreeStore();
export function SegmentViewNode(param) {
    let { type, pagePath, children } = param;
    const clientState = getSegmentTreeClientState();
    const tree = clientState.tree;
    useEffect(()=>{
        if (!tree) {
            return;
        }
        tree.insert({
            type,
            pagePath
        });
    }, [
        type,
        pagePath,
        tree
    ]);
    return children;
}
export function useSegmentTreeClientState() {
    const state = useSyncExternalStore(subscribe, getSnapshot, getServerSnapshot);
    return state;
}
