// next@15.4.0-canary.59/dist/client/components/react-dev-overlay/ui/components/resizer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Resizer", {
    enumerable: true,
    get: function() {
        return Resizer;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _react = require("react");
const Resizer = /*#__PURE__*/ (0, _react.forwardRef)(function Resizer(param, resizerRef) {
    let { children, measure, ...props } = param;
    const ref = (0, _react.useRef)(null);
    const [height, measuring] = useMeasureHeight(ref, measure);
    return /*#__PURE__*/ (0, _jsxruntime.jsx)("div", {
        ...props,
        ref: resizerRef,
        // [x] Don't animate on initial load
        // [x] No duplicate elements
        // [x] Responds to content growth
        style: {
            height: measuring ? 'auto' : height,
            transition: 'height 250ms var(--timing-swift)'
        },
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)("div", {
            ref: ref,
            children: children
        })
    });
});
function useMeasureHeight(ref, measure) {
    const [height, setHeight] = (0, _react.useState)(0);
    const [measuring, setMeasuring] = (0, _react.useState)(true);
    (0, _react.useEffect)(()=>{
        if (!measure) {
            return;
        }
        let timerId;
        const el = ref.current;
        if (!el) {
            return;
        }
        const observer = new ResizeObserver((param)=>{
            let [{ contentRect }] = param;
            clearTimeout(timerId);
            timerId = window.setTimeout(()=>{
                setMeasuring(false);
            }, 100);
            setHeight(contentRect.height);
        });
        observer.observe(el);
        return ()=>observer.disconnect();
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [
        measure
    ]);
    return [
        height,
        measuring
    ];
}

if ((typeof exports.default === 'function' || (typeof exports.default === 'object' && exports.default !== null)) && typeof exports.default.__esModule === 'undefined') {
  Object.defineProperty(exports.default, '__esModule', { value: true });
  Object.assign(exports.default, exports);
  module.exports = exports.default;
}
