// next@15.4.0-canary.59/dist/esm/client/components/react-dev-overlay/ui/container/errors.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useState, useMemo, useRef, Suspense } from 'react';
import { Overlay } from '../components/overlay';
import { RuntimeError } from './runtime-error';
import { getErrorSource } from '../../../../../shared/lib/error-source';
import { HotlinkedText } from '../components/hot-linked-text';
import { PseudoHtmlDiff } from './runtime-error/component-stack-pseudo-html';
import { isConsoleError } from '../../../errors/console-error';
import { extractNextErrorCode } from '../../../../../lib/error-telemetry-utils';
import { ErrorOverlayLayout } from '../components/errors/error-overlay-layout/error-overlay-layout';
import { getHydrationErrorStackInfo, isHydrationError, NEXTJS_HYDRATION_ERROR_LINK } from '../../../react-19-hydration-error';
import { getSquashedHydrationErrorDetails } from '../../pages/hydration-error-state';
import { isRecoverableError } from '../../../../react-client-callbacks/on-recoverable-error';
function isNextjsLink(text) {
    return text.startsWith('https://nextjs.org');
}
function HydrationErrorDescription(param) {
    let { message } = param;
    return /*#__PURE__*/ _jsx(HotlinkedText, {
        text: message,
        matcher: isNextjsLink
    });
}
function GenericErrorDescription(param) {
    let { error } = param;
    const environmentName = 'environmentName' in error ? error.environmentName : '';
    const envPrefix = environmentName ? "[ " + environmentName + " ] " : '';
    // The environment name will be displayed as a label, so remove it
    // from the message (e.g. "[ Server ] hello world" -> "hello world").
    let message = error.message;
    if (message.startsWith(envPrefix)) {
        message = message.slice(envPrefix.length);
    }
    return /*#__PURE__*/ _jsx(_Fragment, {
        children: /*#__PURE__*/ _jsx(HotlinkedText, {
            text: message,
            matcher: isNextjsLink
        })
    });
}
function getErrorType(error) {
    if (isRecoverableError(error)) {
        return "Recoverable " + error.name;
    }
    if (isConsoleError(error)) {
        return "Console " + error.name;
    }
    return "Runtime " + error.name;
}
const noErrorDetails = {
    hydrationWarning: null,
    notes: null,
    reactOutputComponentDiff: null
};
function useErrorDetails(error) {
    return useMemo(()=>{
        if (error === undefined) {
            return noErrorDetails;
        }
        const pagesRouterErrorDetails = getSquashedHydrationErrorDetails(error);
        if (pagesRouterErrorDetails !== null) {
            var _pagesRouterErrorDetails_warning, _pagesRouterErrorDetails_reactOutputComponentDiff;
            return {
                hydrationWarning: (_pagesRouterErrorDetails_warning = pagesRouterErrorDetails.warning) != null ? _pagesRouterErrorDetails_warning : null,
                notes: null,
                reactOutputComponentDiff: (_pagesRouterErrorDetails_reactOutputComponentDiff = pagesRouterErrorDetails.reactOutputComponentDiff) != null ? _pagesRouterErrorDetails_reactOutputComponentDiff : null
            };
        }
        if (!isHydrationError(error)) {
            return noErrorDetails;
        }
        const { message, notes, diff } = getHydrationErrorStackInfo(error);
        if (message === null) {
            return noErrorDetails;
        }
        return {
            hydrationWarning: message,
            notes,
            reactOutputComponentDiff: diff
        };
    }, [
        error
    ]);
}
export function Errors(param) {
    let { runtimeErrors, debugInfo, onClose, ...props } = param;
    const dialogResizerRef = useRef(null);
    const isLoading = useMemo(()=>{
        return runtimeErrors.length < 1;
    }, [
        runtimeErrors.length
    ]);
    const [activeIdx, setActiveIndex] = useState(0);
    const activeError = useMemo(()=>{
        var _runtimeErrors_activeIdx;
        return (_runtimeErrors_activeIdx = runtimeErrors[activeIdx]) != null ? _runtimeErrors_activeIdx : null;
    }, [
        activeIdx,
        runtimeErrors
    ]);
    const errorDetails = useErrorDetails(activeError == null ? void 0 : activeError.error);
    if (isLoading) {
        // TODO: better loading state
        return /*#__PURE__*/ _jsx(Overlay, {});
    }
    if (!activeError) {
        return null;
    }
    const error = activeError.error;
    const isServerError = [
        'server',
        'edge-server'
    ].includes(getErrorSource(error) || '');
    const errorType = getErrorType(error);
    // TOOD: May be better to always treat everything past the first blank line as notes
    // We're currently only special casing hydration error messages.
    const notes = errorDetails.notes;
    const hydrationWarning = errorDetails.hydrationWarning;
    const errorCode = extractNextErrorCode(error);
    const footerMessage = isServerError ? 'This error happened while generating the page. Any console logs will be displayed in the terminal window.' : undefined;
    return /*#__PURE__*/ _jsxs(ErrorOverlayLayout, {
        errorCode: errorCode,
        errorType: errorType,
        errorMessage: hydrationWarning ? /*#__PURE__*/ _jsx(HydrationErrorDescription, {
            message: hydrationWarning
        }) : /*#__PURE__*/ _jsx(GenericErrorDescription, {
            error: error
        }),
        onClose: isServerError ? undefined : onClose,
        debugInfo: debugInfo,
        error: error,
        runtimeErrors: runtimeErrors,
        activeIdx: activeIdx,
        setActiveIndex: setActiveIndex,
        footerMessage: footerMessage,
        dialogResizerRef: dialogResizerRef,
        ...props,
        children: [
            /*#__PURE__*/ _jsxs("div", {
                className: "error-overlay-notes-container",
                children: [
                    notes ? /*#__PURE__*/ _jsx(_Fragment, {
                        children: /*#__PURE__*/ _jsx("p", {
                            id: "nextjs__container_errors__notes",
                            className: "nextjs__container_errors__notes",
                            children: notes
                        })
                    }) : null,
                    hydrationWarning ? /*#__PURE__*/ _jsx("p", {
                        id: "nextjs__container_errors__link",
                        className: "nextjs__container_errors__link",
                        children: /*#__PURE__*/ _jsx(HotlinkedText, {
                            text: "See more info here: " + NEXTJS_HYDRATION_ERROR_LINK
                        })
                    }) : null
                ]
            }),
            errorDetails.reactOutputComponentDiff ? /*#__PURE__*/ _jsx(PseudoHtmlDiff, {
                className: "nextjs__container_errors__component-stack",
                reactOutputComponentDiff: errorDetails.reactOutputComponentDiff || ''
            }) : null,
            /*#__PURE__*/ _jsx(Suspense, {
                fallback: /*#__PURE__*/ _jsx("div", {
                    "data-nextjs-error-suspended": true
                }),
                children: /*#__PURE__*/ _jsx(RuntimeError, {
                    error: activeError,
                    dialogResizerRef: dialogResizerRef
                }, activeError.id.toString())
            })
        ]
    });
}
export const styles = "\n  .nextjs-error-with-static {\n    bottom: calc(16px * 4.5);\n  }\n  p.nextjs__container_errors__link {\n    font-size: var(--size-14);\n  }\n  p.nextjs__container_errors__notes {\n    color: var(--color-stack-notes);\n    font-size: var(--size-14);\n    line-height: 1.5;\n  }\n  .nextjs-container-errors-body > h2:not(:first-child) {\n    margin-top: calc(16px + 8px);\n  }\n  .nextjs-container-errors-body > h2 {\n    color: var(--color-title-color);\n    margin-bottom: 8px;\n    font-size: var(--size-20);\n  }\n  .nextjs-toast-errors-parent {\n    cursor: pointer;\n    transition: transform 0.2s ease;\n  }\n  .nextjs-toast-errors-parent:hover {\n    transform: scale(1.1);\n  }\n  .nextjs-toast-errors {\n    display: flex;\n    align-items: center;\n    justify-content: flex-start;\n  }\n  .nextjs-toast-errors > svg {\n    margin-right: 8px;\n  }\n  .nextjs-toast-hide-button {\n    margin-left: 24px;\n    border: none;\n    background: none;\n    color: var(--color-ansi-bright-white);\n    padding: 0;\n    transition: opacity 0.25s ease;\n    opacity: 0.7;\n  }\n  .nextjs-toast-hide-button:hover {\n    opacity: 1;\n  }\n  .nextjs__container_errors__error_title {\n    display: flex;\n    align-items: center;\n    justify-content: space-between;\n    margin-bottom: 14px;\n  }\n  .error-overlay-notes-container {\n    margin: 8px 2px;\n  }\n  .error-overlay-notes-container p {\n    white-space: pre-wrap;\n  }\n";
